Fix placeholder method error

hoelzro opened the pull request:
Addresses GH #2320 

There are a few things I'd like some discussion on for this PR:

  * I noticed there's an `X::Signature::Placeholder` - should this error be using something like that for its exception?  I understand the documentation for that exception class indicates an explicit signature conflicting with a placeholder, but the class name itself implies a more broad category.
  * I haven't added tests yet - what's the best way to test error messages without them being too brittle?
  * I didn't know that there was such a thing as slurpy placeholders - I don't know what that would look like, so this change might not work with those!

zoffixznet commented:
> what's the best way to test error messages without them being too brittle?

Currently, that goes to https://github.com/rakudo/rakudo/blob/master/t/05-messages/03-errors.t

>  slurpy placeholders - I don't know what that would look 

I believe these are `*@`, `*%`, and `|`, aren't they?

zoffixznet commented on src/Perl6/Actions.nqp:
```diff
@@ -9688,6 +9709,7 @@ class Perl6::Actions is HLL::Actions does STDActions {
             pos_slurpy        => $pos_slurpy,
             named_slurpy      => $named_slurpy,
             placeholder       => $full_name,
+            ast               => $/,
```
Is this a common name used for this purpose?

All I ever seen `ast` used for made stuff, like QAST trees, not for Perl6::Grammar objects. `node` is used for those usually in things like QAST nodes.

hoelzro replied:
Thanks for the suggestion! I'll incorporate that.

hoelzro commented:
Ok, I've added the test and changed `ast` to `node`.

Regarding the slurpy placeholders, looking at the code, it looks like those are `@_` and `%_`.  I ask because my code currently makes the this suggestion:

```
===SORRY!=== Error while compiling /home/rob/projects/rakudo/test.pl
Placeholder variables (eg. $^a) cannot be used in a method.
Please specify an explicit signature, like method z ($a) { ... }
at /home/rob/projects/rakudo/test.pl:6
------>     method z { ⏏$^a }
```

I'm not sure what to suggest if `@_` is used; currently `method z (@) { ... }` is suggested, which is pretty LTA.

hoelzro commented:
I handle `@_` and `%_` now - although it seems that if I try `class { method z { %_ } }` it does _not_ fail to compile, as expected.

hoelzro commented:
Although now, looking at the docs, I can't really see the justification for _why_ `@_` isn't allowed!
